library/std/src/thread/mod.rs

```diff
@@ -1287,12 +1287,31 @@ unsafe impl<'scope, T: Sync> Sync for Packet<'scope, T> {}
 
 impl<'scope, T> Drop for Packet<'scope, T> {
     fn drop(&mut self) {
+        // If this packet was for a thread that ran in a scope, the thread
+        // panicked, and nobody consumed the panic payload, we make sure
+        // the scope function will panic.
+        let unhandled_panic = matches!(self.result.get_mut(), Some(Err(_)));
+        // Drop the result without causing unwinding.
+        // This is only relevant for threads that aren't join()ed, as
+        // join() will take the `result` and set it to None, such that
+        // there is nothing left to drop here.
+        // If this panics, we should handle that, because we're outside the
+        // outermost `catch_unwind` of our thread.
+        // We just abort in that case, since there's nothing else we can do.
+        // (And even if we tried to handle it somehow, we'd also need to handle
+        // the case where the panic payload we get out of it also panics on
+        // drop, and so on. See issue #86027.)
+        if let Err(_) = panic::catch_unwind(panic::AssertUnwindSafe(|| {
+            *self.result.get_mut() = None;
+        })) {
+            rtabort!("thread result panicked on drop");
+        }
```

Comment:
FWIW, doing something like:

```suggestion
        let abort_on_panic = scopeguard::guard((), |()| {
            rt_abort!("thread result panicked on drop");
        });
        *self.result.get_mut() = None;
        let _defused: () = ScopeGuard::into_inner(abort_on_panic);
```

  - (Or some stdlib internal flavor of a drop guard);

I'm personally always fonder of favoring drop guards over `catch_unwind`s, when possible, precisely because it tackles caveats such as #86027 a bit better.

___

But considering such an issue, and more generally, unwind on drop, will be disabled, I agree it may not deserve that much effort

Reply:
I prefer the catch_unwind version, because it more directly represents what we're doing: catching a panic and handling it.